## Assignment on Statistics

In [84]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [85]:
df = pd.read_csv('/content/drive/MyDrive/IIT Palakkad Advanced AI/Assignments/Dataset/StudentsPerformance.csv')
df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group C,some high school,free/reduced,none,0,17,10
1,female,group B,high school,free/reduced,none,8,24,23
2,female,group B,some high school,free/reduced,none,18,32,28
3,female,group B,some college,standard,none,11,38,32
4,female,group C,some college,free/reduced,none,22,39,33


In [86]:
df.dtypes

,0
gender,object
race/ethnicity,object
parental level of education,object
lunch,object
test preparation course,object
math score,int64
reading score,int64
writing score,int64


In [87]:
df.shape

(1000, 8)

In [88]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [89]:
numerical_cols = [column for column in df.columns if df[column].dtype != 'O']
numerical_cols

['math score', 'reading score', 'writing score']

### 1. Calculate the mean, median, and mode for the math, reading, and writing scores.

Mean and median of math, reading and writing score

In [90]:
df.mean(numeric_only=True)

,0
math score,66.001
reading score,69.195
writing score,68.119


In [91]:
df.median(numeric_only=True)

,0
math score,66.0
reading score,70.0
writing score,69.0


mode for math, reading and writing scores

In [92]:
for col in numerical_cols:
  print(f'Mode of {col} = {df[col].mode().tolist()}')

Mode of math score = [65]
Mode of reading score = [72]
Mode of writing score = [74]


### 2. Determine the range, variance, and standard deviation for each score category.

Range of math, reading and writing scores

In [93]:
for col in numerical_cols:
  print(f'Range of {col} = {df[col].max() - df[col].min()}')

Range of math score = 100
Range of reading score = 83
Range of writing score = 90


Variance of math, reading and writing scores

In [94]:
df.var(numeric_only=True)

,0
math score,242.411410
reading score,216.279254
writing score,231.842682


Standard Devaition of math, reading and writing scores

In [95]:
df.std(numeric_only=True)

,0
math score,15.569567
reading score,14.706436
writing score,15.226381


### 3. Identify the mean, median, and mode of math scores for students with different parental education levels.

In [96]:
df.groupby('parental level of education').mean(numeric_only=True).rename(columns=lambda col: f'{col} mean')

,math score mean,reading score mean,writing score mean
parental level of education,,,
associate's degree,67.977477,71.018018,70.031532
bachelor's degree,69.288136,73.000000,73.381356
high school,61.821429,64.602041,62.403061
master's degree,70.254237,75.949153,75.677966
some college,67.128319,69.566372,69.035398
some high school,63.134078,66.759777,64.888268


In [97]:
df.groupby('parental level of education').median(numeric_only=True).rename(columns=lambda col: f'{col} median')

,math score median,reading score median,writing score median
parental level of education,,,
associate's degree,67.0,73.0,71.5
bachelor's degree,68.0,73.0,74.0
high school,63.0,66.0,64.0
master's degree,73.0,76.0,75.0
some college,68.0,71.0,70.0
some high school,65.0,67.0,66.0


In [98]:
df.groupby('parental level of education').agg(lambda x: x.mode().tolist()).drop(columns=['gender', 'race/ethnicity', 'lunch', 'test preparation course']).rename(columns=lambda col: f'{col} mode')

,math score mode,reading score mode,writing score mode
parental level of education,,,
associate's degree,[65],"[76, 77]",[73]
bachelor's degree,"[63, 65, 66]",[100],"[81, 100]"
high school,"[57, 66]","[52, 72]",[68]
master's degree,"[79, 81]",[81],"[75, 84, 100]"
some college,[69],"[64, 74, 78]",[70]
some high school,"[59, 65]","[67, 84]",[78]


### 4. Compare the mean scores of students who took the test preparation course versus those who didn’t.

In [99]:
df.groupby('test preparation course').mean(numeric_only=True)

,math score,reading score,writing score
test preparation course,,,
completed,69.96648,74.175978,74.684358
none,63.78972,66.417445,64.457944


Students who completed the test preparation course scored slightly higher in math on average compared to those who did not.

For reading and writing, the difference was more notable, students who completed the course performed considerably higher on average than those who skipped it.

### 5. Calculate the interquartile range (IQR) for each score category.

In [100]:
df.columns

Index(['gender', 'race/ethnicity', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')

In [101]:
for col in numerical_cols:
  Q1 = df[col].quantile(0.25)
  Q3 = df[col].quantile(0.75)
  IQR = Q3 - Q1
  print(f'IQR of {col} = {IQR}')

IQR of math score = 20.0
IQR of reading score = 20.25
IQR of writing score = 21.0


### 6. Determine the standard deviation for math scores by lunch type (free/reduced vs. standard).

In [102]:
df.groupby('lunch')['math score'].std().rename('math score standard deviation')

,math score standard deviation
lunch,
free/reduced,15.882442
standard,13.804560


### 7. Find the correlation coefficient between math and reading scores, and then between reading and writing scores.

In [103]:
df.corr(numeric_only=True)

,math score,reading score,writing score
math score,1.000000,0.814937,0.797158
reading score,0.814937,1.000000,0.953178
writing score,0.797158,0.953178,1.000000


### 8. Interpret the strength and direction of the correlation. What might this suggest about the relationship between different scores?

**Math & Reading (0.815)**

- math and reading scores are highly correlated (has strong correlation)
- slightly weaker than reading-writing
- positive correlation


**Reading & Writing (0.953)**

- reading score and writing score is highly correlated (strong positive correlation).
- students who score high in reading tend to also score high in reading

### 9. Calculate the covariance between math and reading scores. How does it compare with the correlation? What does it suggest?

In [104]:
df[['math score', 'reading score']].cov()

,math score,reading score
math score,242.411410,186.598403
reading score,186.598403,216.279254


In [105]:
df[['math score', 'reading score']].corr()

,math score,reading score
math score,1.000000,0.814937
reading score,0.814937,1.000000


Covariance between math and reading scores = 186.60

- The positive value confirms that math and reading scores move in the same direction
- students who score higher in math tend to score higher in reading too.




Covariance (186.6)


- Shows direction of relationship
- Positive => both variables increase together
- Value depends on the scale (so 186 doesn’t mean “very strong” by itself)


Correlation (0.814937)


- Shows strength + direction
- Standardized between -1 and 1
- Easier to interpret

### 10. Z-Test: Assume the national average math score is 50. Use a one-sample z-test to test if this sample's average math score is significantly different from the national average.

In [106]:
# One-Sample Z-Test: Math Score vs National Average

# We test whether the sample mean math score
# is significantly different from the national average (50)

import numpy as np
from scipy.stats import norm

In [107]:
# Step 1: Define population mean (national average)
mu0 = 50

# Sample mean (x̄)
x_bar = df['math score'].mean()

# Sample standard deviation (s)
s = df['math score'].std()

# Sample size (n)
n = df['math score'].count()

print("Sample mean:", x_bar)
print("Sample std:", s)
print("Sample size:", n)

Sample mean: 66.001
Sample std: 15.569566802271998
Sample size: 1000


In [108]:
# Step 3: Compute Z-score

# Formula:
# z = (x̄ - μ0) / (s / sqrt(n))

z = (x_bar - mu0) / (s / np.sqrt(n))

print("Z-score:", z)

Z-score: 32.49904476017308


In [109]:
# Step 4: Compute p-value (two-tailed test)

# We multiply by 2 because it's a two-sided test
p_value = 2 * (1 - norm.cdf(abs(z)))

print("P-value:", p_value)

P-value: 0.0


In [110]:
# Step 5: Decision

alpha = 0.05  # significance level

if p_value < alpha:
    print("Reject H0: Mean is significantly different from 50")
else:
    print("Fail to reject H0: No significant difference from 50")

Reject H0: Mean is significantly different from 50


### 11. T-Test: Compare the reading scores between students with different lunch types (free/reduced vs. standard) to test if there is a significant difference.

In [111]:
# Independent Two-Sample T-Test

import numpy as np
from scipy.stats import ttest_ind

In [112]:
# Step 1: Split the data into two groups

# Students with free/reduced lunch
group_free = df[df['lunch'] == 'free/reduced']['reading score']

# Students with standard lunch
group_standard = df[df['lunch'] == 'standard']['reading score']

print("Free/Reduced count:", len(group_free))
print("Standard count:", len(group_standard))

Free/Reduced count: 355
Standard count: 645


In [113]:
# Step 2: Compute basic statistics (optional but useful)

print("Free/Reduced mean:", group_free.mean())
print("Standard mean:", group_standard.mean())

Free/Reduced mean: 64.6225352112676
Standard mean: 71.71162790697674


In [114]:
# Step 3: Perform Independent T-Test

# H0: No difference in reading scores between groups
# H1: There is a difference

t_stat, p_value = ttest_ind(group_free, group_standard)

print("T-statistic:", t_stat)
print("P-value:", p_value)

T-statistic: -7.492798943109684
P-value: 1.4829967008771313e-13


In [115]:
# Step 4: Decision Rule

alpha = 0.05  # significance level

if p_value < alpha:
    print("Reject H0: Significant difference in reading scores")
else:
    print("Fail to reject H0: No significant difference")

Reject H0: Significant difference in reading scores


### 12. F-Test: Check if the variances of math scores differ significantly between students with and without test preparation.

In [116]:
# F-Test for Variance Comparison

import numpy as np
from scipy.stats import f

In [117]:
# Step 1: Split the data into two groups

# Students who completed test preparation
prep_yes = df[df['test preparation course'] == 'completed']['math score']

# Students who did NOT complete test preparation
prep_no = df[df['test preparation course'] == 'none']['math score']

print("Completed prep count:", len(prep_yes))
print("No prep count:", len(prep_no))

Completed prep count: 358
No prep count: 642


In [118]:
# Step 2: Compute sample variances

var_yes = np.var(prep_yes, ddof=1)  # sample variance
var_no = np.var(prep_no, ddof=1)

print("Variance (completed):", var_yes)
print("Variance (none):", var_no)

Variance (completed): 210.8840273539583
Variance (none): 246.66866169973943


In [119]:
# Step 3: Compute F-statistic

# Always put larger variance in numerator
F = var_yes / var_no if var_yes > var_no else var_no / var_yes

print("F-statistic:", F)

F-statistic: 1.1696886900102605


In [120]:
# Step 4: Degrees of freedom

df1 = len(prep_yes) - 1
df2 = len(prep_no) - 1

print("df1:", df1)
print("df2:", df2)

df1: 357
df2: 641


In [121]:
# Step 5: Compute p-value

# We use survival function (1 - cdf)
p_value = 1 - f.cdf(F, df1, df2)

# Two-tailed adjustment
p_value *= 2

print("P-value:", p_value)

P-value: 0.08951037679012552


In [122]:
# Step 6: Decision Rule

alpha = 0.05  # significance level

if p_value < alpha:
    print("Reject H0: Variances are significantly different")
else:
    print("Fail to reject H0: No significant difference in variances")

Fail to reject H0: No significant difference in variances
